keps/0010-dynamic-plugin-configuration.md

```diff
@@ -0,0 +1,221 @@
+    - [Story 5](#story-5)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+    - [Notes](#notes)
+    - [Caveats](#caveats)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently the configuration for the `KongPlugin` and `KongClusterPlugin` resources
+is generally static, meaning that it is not possible to derive configuration values from
+common configuration resources defined on same kubernetes cluster. This leads to a lot
+duplication of hard coded values in numerous places making the deployment process
+difficult to manage. While it is possible to load configuration from a secret using
+the `configFrom` parameter, the entire configuration must be contained with in the
+secret which is effectively the same problem.
+
+It would be advantageous, and more in line with common Kubernetes behavior to be able
+to derive individual values of plugin configuration from data already stored on the same
+Kubernetes cluster. This proposal aims to define a change to the `KongPlugin` and `KongClusterPlugin`
+Custom Resource Definitions to facilitate this behavior.
+
+## Motivation
+
+Allow `KongPlugin` and `KongClusterPlugin` resources to read configuration values from Kubernetes,
+in addition to literal values.
+
+Continue to allow `KongPlugin` and `KongClusterPlugin` resource to accept literal input values.
+
+Reduce the amount of duplication of configuration value in the simple case. That is,
+when using standard kubernetes objects and with out assumed templating systems.
+
+### Goals
+
+* Provide ability to read individual configuration values from `ConfigMap` resources
+* Provide ability to read individual configuration values from `Secret` resources
+* Provide ability to read individual configuration values from arbitrary or generic resource types (other custom resource types)
+
+### Non-Goals
+
+* Introduction of a new Custom Resource or different Plugin Resource
+
+## Proposal
+
+In an effort to stay in line with current Kubernetes resource patterns it would be desirable
+to mimic the way pods define environment variables using an array of objects where
+the objects contain a `name` and one of `value` or `valueFrom`. `valueFrom`
+would further supports `configMapKeyRef`, `secretKeyRef` or `genericKeyRef`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a operator I want certain plugin configuration values to be dynamic in addition
+to statically definable
+
+#### Story 2
+
+As a operator I want to be able to derive a literal value using value interpolation referencing
+other values
+
+#### Story 3
+
+As an operator I want values stored in `ConfigMap` or `Secret` resources
+to be the source of truth for various resources deployed to my Kubernetes cluster
+
+#### Story 4
+
+As an operator I expect changes in `ConfigMap` or `Secret` data values to be reflected
+in the resource and deployments that depend on them with out having to restart critical parts of the system
```

Comment:
This is a particularly interesting user story. It would seem we would want to add something to the goals for this. Perhaps something like:

- the `KongPlugin` controller should react to updates to `ConfigMap` and `Secret` objects referenced by managed plugins to re-apply the configuration to all attached `KongPlugins`

Does that resonate with you?

Also this has some implications in terms of danger around operators changing commonly used `ConfigMaps` and `Secrets` as those changes could cascade to any number of otherwise configured and operating plugins. Any thoughts on some of the potential hazards?

Reply:
Right. one of the very nice things about the plugins now being basically static + the controller loops is they are effectively Hot reloadable. updating the plugin - wait 5 seconds.

In this model that breaks, and I would have to restart the kong deployments - which has implications.

Stress on the kubernetes API server is a real thing. we use a controller the just runs on an interval and that has been far worse than event driven updates. Another problem is going to be resources that get deleted may have a negative cascade / deck of cards effect. You might be able to prevent that with a finalizer + child resources (maybe?). 

Its probably not save to assume everyone does this, but we keep config maps pretty tight to use case - all redis config is in 1  `redis` for example. Try to avoid dumping ground config map that would have high frequency updates.

As an alternative, I've seen other controllers use annotations to allow opt-in to watching other resources. not sure I'd like that in this case.

Reply:
I don't think the cascade of changing 1 secret down to many plugins is a bad thing, I'd expect that.

Reply:
> I don't think the cascade of changing 1 secret down to many plugins is a bad thing, I'd expect that.

Indeed. I certainly agree it's an obvious implication, but I was more wondering if you feel like there's anything we can do to ease the pain of unintended consequence: e.g. if someone were to update a `ConfigMap` with an accidental _bad configuration_ (think similar to when ISPs accidentally mess up their BGP configurations, but certainly on a much smaller scale) are there any things we can do to reduce or help in these situations :thinking: I'm kinda thinking it's a _no_, and we just have to consider the risk part of the cost of offering this kind of solution.

On the other hand, for the purposes of resolving this comment I would just like to see us reflect this user story as a goal, perhaps similar to what I suggested above, as it seems this is a fairly important design goal.

Reply:
well there is some validation at the plugin level with the admission controller which should prevent things from being overtly invalid. outside of that, I think that is mostly contextual - not much you can control at that point

Reply:
Added to goals as well